flair/models/biomedical_entity_linking.py

```diff
@@ -0,0 +1,988 @@
+                        )
+                    else:
+                        index = np.where(ids == sparse_id)[0][0]
+                        distances[index] = (
+                            sparse_weight * sparse_distance
+                        ) + distances[index]
+
+                sorted_indizes = np.argsort(-distances)
+                ids = ids[sorted_indizes][:topk]
+                distances = distances[sorted_indizes][:topk]
+                hybrid_ids.append(ids.tolist())
+                hybrid_scores.append(distances.tolist())
+
+            return [
+                np.append(self.dictionary[ind], score)
+                for ind, score in zip(hybrid_ids, hybrid_scores)
+            ]
+        # use only dense embeds
+        else:
+            dense_distances, dense_ids = self.dense_dictionary_index.search(
+                x=mention_dense_embeds, k=topk
+            )
+            return [
+                np.append(self.dictionary[ind], score)
+                for ind, score in zip(dense_ids, dense_distances)
+            ]
+
+    def embed_dictionary(
+        self,
+        model_name_or_path: str,
+        dictionary_name_or_path: str,
+        mean_centering: bool,
+        batch_size: int,
+    ):
+        # check for embedded dictionary in cache
+        dictionary_name = os.path.splitext(os.path.basename(dictionary_name_or_path))[0]
+        cache_folder = os.path.join(flair.cache_root, "datasets")
+        file_name = f"bio_nen_{model_name_or_path.split('/')[-1]}_{dictionary_name}"
+        cached_dictionary_path = os.path.join(
+            cache_folder,
+            f"{file_name}.pk",
+        )
+
+        # If exists, load the cached dictionary indices
+        if os.path.exists(cached_dictionary_path):
+            self.load_cached_dictionary(cached_dictionary_path)
+
+        # else, load and embed
+        else:
+            self.load_dictionary(dictionary_name_or_path)
+
+            # embed dictionary
+            dictionary_names = [row[0] for row in self.dictionary]
+
+            # create dense and sparse embeddings
+            if self.use_sparse_embeds:
+                self.dict_dense_embeds = self.embed_dense(
+                    names=dictionary_names, show_progress=True
+                )
+                self.dict_sparse_embeds = self.embed_sparse(
+                    names=dictionary_names, show_progress=True, batch_size=batch_size
+                )
+
+            # create only dense embeddings
+            else:
+                self.dict_dense_embeds = self.embed_dense(
+                    names=dictionary_names, show_progress=True
+                )
+                self.dict_sparse_embeds = None
+
+            # build dense index
+            dimension = self.dict_dense_embeds.shape[1]
+            # to use cosine similarity, we normalize the vectors and then use inner product
+            faiss.normalize_L2(self.dict_dense_embeds)
```

Comment:
Ok, we should the have a `use_cosine_similarity` arg for `BiEncoder` to decide which metric to use. However, from my experience IP works just fine